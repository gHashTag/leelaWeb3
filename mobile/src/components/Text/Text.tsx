import React, { memo } from 'react'

import {
  Platform,
  Text as RNText,
  StyleProp,
  StyleSheet,
  TextProps,
  TextStyle,
  useColorScheme,
} from 'react-native'

import { gray } from 'cons'
import { ms, s } from 'react-native-size-matters'

export type hT = 'h0' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5'

export interface Icolors {
  dark: string
  light: string
}

export interface TxtT extends TextProps {
  title: string
  h?: hT
  colors?: Icolors
  oneColor?: string
  numberOfLines?: number
  textStyle?: StyleProp<TextStyle>
  testID?: string
  onPress?: () => void
}

const Text = memo<TxtT>(
  ({
    h,
    colors,
    title,
    oneColor = gray,
    numberOfLines,
    textStyle,
    testID = 'text',
    onPress,
    ...textProps
  }) => {
    const scheme = useColorScheme()
    const isDark = scheme === 'dark'

    let curColor = oneColor
    if (colors) {
      curColor = isDark ? colors.dark : colors.light
    }

    let hStyle: TextStyle | undefined
    if (h && textStyles[h]) {
      // @ts-ignore
      hStyle = { ...textStyles[h], color: curColor }
    }

    const mergedStyles: StyleProp<TextStyle> = StyleSheet.flatten([
      hStyle,
      textStyle,
    ])

    return (
      <RNText
        style={mergedStyles}
        {...textProps}
        testID={testID}
        ellipsizeMode="tail"
        numberOfLines={numberOfLines}
        onPress={onPress}
      >
        {title}
      </RNText>
    )
  },
)

export const textStyles = StyleSheet.create({
  h0: {
    fontFamily: Platform.OS === 'ios' ? 'Etna' : 'etna-free-font',
    fontSize: Platform.OS === 'ios' ? ms(95, 0.5) : ms(95, 0.3),
    textShadowOffset: { width: 1, height: 1 },
    textShadowRadius: 1,
  },
  h1: {
    fontFamily: 'mont',
    fontSize: Platform.OS === 'ios' ? ms(35, 0.3) : ms(35, 0.6),
    fontWeight: 'bold',
    textShadowOffset: { width: 1, height: 1 },
    textShadowRadius: 1,
  },
  h2: {
    fontFamily: 'mont',
    fontSize: Platform.OS === 'ios' ? s(20) : s(20),
    fontWeight: 'bold',
    lineHeight: 20,
  },
  h3: {
    fontFamily: 'mont',
    fontSize: Platform.OS === 'ios' ? ms(18, 0.6) : ms(18, 0.6),
    lineHeight: 20,
    textShadowOffset: { width: 1, height: 1 },
    textShadowRadius: 1,
  },
  h4: {
    fontFamily: Platform.OS === 'ios' ? 'mont' : 'mont',
    fontSize: Platform.OS === 'ios' ? s(15) : s(15),
    lineHeight: 18,
  },
  h5: {
    fontFamily: 'mont',
    fontSize: Platform.OS === 'ios' ? ms(12, 0.8) : s(15),
  },
})

export { Text }
