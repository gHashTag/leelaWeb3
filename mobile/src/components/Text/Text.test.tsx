import React from 'react'

import { TextStyle } from 'react-native'

import { fireEvent, render } from '@testing-library/react-native'
import renderer from 'react-test-renderer'

import { Text, hT, textStyles } from './Text'

describe('Text Component', () => {
  it('Text component snapshot', () => {
    const component = renderer.create(
      <Text title="Snapshot Test" testID="snapshot-text" />,
    )

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with default props', () => {
    const { getByText } = render(
      <Text testID="text-component" title="Test Title" />,
    )
    expect(getByText('Test Title')).toBeTruthy()
  })

  it('applies the correct style based on the "h" prop', () => {
    const styles: hT[] = ['h0', 'h1', 'h2', 'h3', 'h4', 'h5']
    styles.forEach((style) => {
      const { getByTestId } = render(
        <Text testID="text-component" title="Test Title" h={style} />,
      )
      const receivedStyle = getByTestId('text-component').props.style
      expect(receivedStyle).toEqual(expect.objectContaining(textStyles[style]))
    })
  })

  it('should call the onPress function when pressed', () => {
    const onPressMock = jest.fn()
    const { getByTestId } = render(
      <Text
        title="Clickable Text"
        onPress={onPressMock}
        testID="clickable-text"
      />,
    )
    const clickableText = getByTestId('clickable-text')
    fireEvent.press(clickableText)
    expect(onPressMock).toHaveBeenCalled()
  })

  it('displays the correct title', () => {
    const title = 'Test Title'
    const { getByTestId } = render(
      <Text testID="text-component" title={title} />,
    )
    const textComponent = getByTestId('text-component')
    expect(textComponent.props.children).toBe(title)
  })

  it('displays truncated text when numberOfLines is set', () => {
    const longText = 'This is a very long text that should be truncated'
    const { getByTestId } = render(
      <Text testID="text-component" title={longText} numberOfLines={1} />,
    )
    const textComponent = getByTestId('text-component')
    expect(textComponent.props.children).toBe(longText)
  })

  it('applies additional custom styles', () => {
    const customStyle: TextStyle = { fontSize: 20, fontWeight: 'bold' }
    const { getByTestId } = render(
      <Text
        testID="text-component"
        title="Test Title"
        textStyle={customStyle}
      />,
    )
    const textComponent = getByTestId('text-component')
    const receivedStyle = textComponent.props.style
    expect(receivedStyle).toEqual(expect.objectContaining(customStyle))
  })
})
