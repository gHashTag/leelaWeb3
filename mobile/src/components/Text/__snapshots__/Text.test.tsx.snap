// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component Text component snapshot 1`] = `
<Text
  ellipsizeMode="tail"
  style={{}}
  testID="snapshot-text"
>
  Snapshot Test
</Text>
`;
